services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java

```diff
@@ -267,34 +189,12 @@ private void checkGrantType() {
             throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_REQUEST, "Missing form parameter: " + OIDCLoginProtocol.GRANT_TYPE_PARAM, Response.Status.BAD_REQUEST);
         }
 
-        if (grantType.equals(OAuth2Constants.AUTHORIZATION_CODE)) {
-            event.event(EventType.CODE_TO_TOKEN);
-            action = Action.AUTHORIZATION_CODE;
-        } else if (grantType.equals(OAuth2Constants.REFRESH_TOKEN)) {
-            event.event(EventType.REFRESH_TOKEN);
-            action = Action.REFRESH_TOKEN;
-        } else if (grantType.equals(OAuth2Constants.PASSWORD)) {
-            event.event(EventType.LOGIN);
-            action = Action.PASSWORD;
-        } else if (grantType.equals(OAuth2Constants.CLIENT_CREDENTIALS)) {
-            event.event(EventType.CLIENT_LOGIN);
-            action = Action.CLIENT_CREDENTIALS;
-        } else if (grantType.equals(OAuth2Constants.TOKEN_EXCHANGE_GRANT_TYPE)) {
-            event.event(EventType.TOKEN_EXCHANGE);
-            action = Action.TOKEN_EXCHANGE;
-        } else if (grantType.equals(OAuth2Constants.UMA_GRANT_TYPE)) {
-            event.event(EventType.PERMISSION_TOKEN);
-            action = Action.PERMISSION;
-        } else if (grantType.equals(OAuth2Constants.DEVICE_CODE_GRANT_TYPE)) {
-            if (!Profile.isFeatureEnabled(Profile.Feature.DEVICE_FLOW)) {
-                throw newUnsupportedGrantTypeException();
-            }
-            event.event(EventType.OAUTH2_DEVICE_CODE_TO_TOKEN);
-            action = Action.OAUTH2_DEVICE_CODE;
-        } else if (grantType.equals(OAuth2Constants.CIBA_GRANT_TYPE)) {
-            event.event(EventType.AUTHREQID_TO_TOKEN);
-            action = Action.CIBA;
-        } else {
+        grant = OAuth2GrantManager.resolve(session, grantType).orElseThrow(() -> newUnsupportedGrantTypeException());
+
+        event.event(grant.getEventType());
+
+        if (grantType.equals(OAuth2Constants.DEVICE_CODE_GRANT_TYPE)
```

Comment:
Any reason to not rely on `EnvironmentDependentProviderFactory` and avoid this check?

Reply:
Good point. Updated the code to use `EnvironmentDependentProviderFactory` for CIBA, Device Code and Token Exchange.

The side effect of this will be Keycloak returning `400 Bad Request` instead of `501 Not Implemented` for disabled grant types, as we will no longer be calling `ProfileHelper.requireFeature()`.